Created math-related methods which return Vector3 instances from Position objects

TwistedAsylumMC opened the pull request:
Previously, Position::add() and several other methods would return a Vector3 instance as they are from the parent (Vector3) class

## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
This pull request implements a method multiple methods to the Position class which were inherited from the parent Vector3 class, but returning an instance of Position instead.

### Relevant issues
<!-- List relevant issues here -->
<!--

* Fixes #1
* Fixes #2

-->

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
Created ``Position::add()``
Created ``Posisition::subtract()``
Created ``Posisition::multiply()``
Created ``Posisition::divide()``
Created ``Posisition::ceil()``
Created ``Posisition::floor()``
Created ``Posisition::round()``
Created ``Posisition::abs()``

### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->
None :)

## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
There is no backwards compatability changes, only people currently using Position::add() and the other methods will now receive a Position instance, but this will not affect anything.

## Follow-up
<!-- Suggest any actions to be done before/after merging this pull request -->
<!--

Requires translations:

| Name | Value in eng.ini |
| :--: | :---: |
| `foo.bar` | `Foo bar` |

-->

## Tests
<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->
```php
$position = new Position(0, 0, 0, Server::getInstance()->getLevelByName("Test"));
$newPosition = $position->add(1, 1, 1);
var_dump($newPosition->__toString());
```
Output: ``string(32) "Position(level=Test,x=1,y=1,z=1)"``

TwistedAsylumMC committed:
Created Position::add() method

Previously, Position::add() would return a Vector3 instance as it is from the parent (Vector3) class

Sandertv commented:
It seems rather inconsistent to do this only for the `add()` method, whereas methods such as `subtract()` still return Vector3s.

TwistedAsylumMC commented:
Oh oops, will add that now. Forgot about that haha

TwistedAsylumMC committed:
Created other methods which return Vector3 instances

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
+		}
+	}
+	
+	public function multiply(float $number) : Position{
```
I don't see the point of multiplying a `Position`. A `Position`, by definition, is a positional vector. There is no point of scalar-multiplying a positional vector.

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
+			new Position((int) round($this->x, $precision, $mode), (int) round($this->y, $precision, $mode), (int) round($this->z, $precision, $mode), $this->level);
+	}
+	
+	public function abs() : Position{
```
Same as multiply &mdash; there is no reason to take the absolute value of a positional vector. The origin or the octant doesn't really mean anything in the positional sense.

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
+	 * @return Vector3
+	 */
+	public function add($x, $y = 0, $z = 0) : Position{
+		if($x instanceof Position){
```
A Position object is a positional vector by definition. There is no reason to add two positional vectors.

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
+		if($x instanceof Position){
+			return new Position($this->x + $x->x, $this->y + $x->y, $this->z + $x->z);
+		}else{
+			return new Position($this->x + $x, $this->y + $y, $this->z + $z);
```
Shouldn't the level field be carried over?

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
+	 */
+	public function subtract($x = 0, $y = 0, $z = 0) : Position{
+		if($x instanceof Position){
+			return $this->add(-$x->x, -$x->y, -$x->z);
```
The difference between two positions is actually a relative vector. A relative vector should not be represented with the `Position` class.

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
+		return new Position($this->x * $number, $this->y * $number, $this->z * $number, $this->level);
+	}
+	
+	public function divide(float $number) : Position{
```
Same problem with multiply().

SOF3 reviewed (changes requested):
We have been looking to remove the inheritance relationship between Position and Vector3, preferring composition over inheritance. This pull request is actually one reason why it is necessary to separate &mdash; inheritance is not an appropriate approach between these two classes.
Merging this pull request implies that every future method in Vector3 has to be overridden by an equivalent method in Position, otherwise leading to inconsistency. This does not seem to be a sustainable approach.
In addition, manipulating vectors is usually something mathematical, and should not be mixed up with the semantic role of the Position class.

SOF3 commented on src/pocketmine/level/Position.php:
```diff
@@ -89,6 +89,62 @@ public function setLevel(Level $level = null){
 	}
 
+	/**
+	 * @param Position|int $x
```
This is backwards-incompatible, disallowing passing a `Vector3` to `Position->add()`.
This also violates covariance, because this method is expected to work but conditionally doesn't:
```php
function foo(Vector3 $a) : Vector3 {
    return $a->add(new Vector3(1, 2, 3));
}
```
Calling `foo` with a `Position` object leads to a crash (trying to add `$this->x` to a `Vector` object), while `foo` did not even need to depend on pmmp.

SOF3 commented:
> ## Tests
> 
> ```
> $position = new Position(0, 0, 0, Server::getInstance()->getLevelByName("Test"));
> $newPosition = $position->add(1, 1, 1);
> var_dump($newPosition->__toString());
> ```
> 
> Output: `string(32) "Position(level=Test,x=1,y=1,z=1)"`

Is this a real test, or did you just made this up? I don't see a level being passed to `$newPosition` in any way.

TwistedAsylumMC committed:
Removed some methods, fixed some problems

Removed:
 - multiply()
 - divide()
 - abs()
Fixed:
 - backwards-incompatible problem with add() and subtract()
 - fixed some phpdocs
 - fixed a problem where add() and subtract() would not carry over the level

Eren5960 commented:
`Position::fromObject($position->add(1, 1, 1), $level);` is not it enough?

dktapps commented:
This has been discussed on and off for years, and we've always decided against it because of the amount of extra complexity it creates for any descendent class. The benefit is not worth the extra mess it creates. `Position` should never have extended `Vector3` to begin with.
